src/Array.mo

```diff
@@ -38,6 +39,64 @@ module {
       };
     };
   };
+
+  /// Sorts the given array according to the `cmp` function.
+  /// This is a _stable_ sort.
+  ///
+  /// ```motoko
+  /// import Array "mo:base/Array";
+  /// import Nat "mo:base/Nat";
+  /// let xs = [4, 2, 6, 1, 5];
+  /// assert(Array.sort(xs, Nat.compare) == [1, 2, 4, 5, 6])
+  /// ```
+  public func sort<A>(xs : [A], cmp : (A, A) -> Order.Order) : [A] {
+    let tmp : [var A] = thaw(xs);
+    sortInPlace(tmp, cmp);
+    freeze(tmp)
+  };
+
+  public func sortInPlace<A>(xs : [var A], cmp : (A, A) -> Order.Order) {
+    if (xs.size() == 0) return;
+    let aux : [var A] = tabulateVar<A>(xs.size(), func i { xs[i] });
+
+    func merge(a : [var A], lo : Nat, mid : Nat, hi : Nat) {
```

Comment:
Is `a` always `xs`? then why pass it around?

Maybe because you don’t want closures, but direct function calls? That’s laudable, maybe even worthwhile, but then you should also pass `xs` and `cmp` around, so that these functions are closed. (You can defined closed functions locally, the backend will treat them like top-level functions.)

Reply:
I appreciate you trying to justify my dumb code, but it was really just an oversight. The Java algorithm has a class member for `aux` but passes the array around, and I copied it without thinking too much.

I'd prefer to keep the code simple for now.